core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,323 @@
+                            }
+
+                            final Header header = method.getAnnotation(Header.class);
+                            if (header != null) {
+                                paramList.add(new HeaderMethodParamInfo(method, header));
+                            }
+                        }
+                );
+
+        if (paramList.isEmpty()) {
+            // there is no fields annotated by @Param or @Header
+            // we should not handle this result type
+            return null;
+        }
+
+        return new BeanParamInfo(expectedResultType, paramList);
+    }
+
+    @Override
+    public Object convertRequest(final ServiceRequestContext ctx,
+                                 final AggregatedHttpMessage request,
+                                 final Class<?> expectedResultType) throws Exception {
+        final BeanParamInfo beanParamInfo = beanParamInfoCache.get(expectedResultType);
+
+        if (beanParamInfo == null) {
+            // no BeanParamInfo for this result type, we should not handle it
+            return RequestConverterFunction.fallthrough();
+        }
+
+        final Object requestBean = beanParamInfo.createInstance();
+        final MutableValueHolder<HttpParameters> httpParams = MutableValueHolder.ofEmpty();
+
+        beanParamInfo.paramList
+                .forEach(
+                        paramInfo -> paramInfo.setValue(ctx, request, httpParams, requestBean)
+                );
+
+        return requestBean;
+    }
+
+    @FunctionalInterface
+    private interface ParamValueRetriever {
+        static ParamValueRetriever fromParam(final String paramName) {
+            return (ctx, request, httpParams) -> {
+                String paramValueStr = ctx.pathParam(paramName);
+                if (paramValueStr == null) {
+                    paramValueStr =
+                            httpParams.computeIfAbsent(
+                                    () -> httpParametersOf(ctx, request.headers(), request)
+                            ).get(paramName);
+                }
+
+                return paramValueStr;
+            };
+        }
+
+        static ParamValueRetriever fromHeader(final String headerName) {
+            return (ctx, request, httpParams) -> request.headers().get(AsciiString.of(headerName));
+        }
+
+        String getParamValue(ServiceRequestContext ctx,
+                             AggregatedHttpMessage request,
+                             MutableValueHolder<HttpParameters> httpParams);
+    }
+
+    private static final class BeanParamInfo {
+        private final Class<?> beanClass;
+
+        private final Constructor<?> constructor;
+
+        private final List<ParamInfo> paramList;
+
+        private BeanParamInfo(final Class<?> beanClass, final List<ParamInfo> paramList) {
+            this.beanClass = beanClass;
+            this.paramList = paramList;
+
+            try {
+                constructor = beanClass.getDeclaredConstructor();
```

Comment:
I guess users may want to avoid using the default constructor but inject the constructor parameter. Could you add a `TODO` item like this:

```java
// TODO: Support injecting constructor parameters.
```

Or, if you feel like, you can implement that in this PR as well.

Reply:
I will implement it later.

Reply:
Done.
now `@Param`/`@Header` can be used to bean constructor and constructor params